### PR TITLE
add targetting specific subnets for transit gateway attachment

### PR DESCRIPTION
**Pull request by bdehri**
This PR:

- Changes operator to target specific subnets with tag `subnet.giantswarm.io/tgw-attachments: "true"` to attach transit gateway.  Previous behavior stays as fallback method.

### Checklist

- [x] Update changelog in CHANGELOG.md.
- [x] Make sure `values.yaml` and `values.schema.json` are valid.

### CODE REVIEW

File: pkg/registrar/transitgateway.go [hunk 1]
```diff
@@ -771,6 +775,36 @@ func (r *TransitGateway) removeFromPrefixList(ctx context.Context, awsCluster *c
 	return nil
 }
 
+func (r *TransitGateway) getTGWGAttachmentSubnetsOrDefault(ctx context.Context, awsCluster *capa.AWSCluster) ([]string, error) {
+	result := make([]string, 0)
+	output, err := r.transitGatewayClient.DescribeSubnets(ctx, &ec2.DescribeSubnetsInput{
+		Filters: []types.Filter{
+			{Name: aws.String("tag:" + capa.NameKubernetesAWSCloudProviderPrefix + awsCluster.Name), Values: []string{"owned", "shared"}},
+			{Name: aws.String("tag:" + subnetTGWAttachementslabel), Values: []string{"true"}},
```
Comment (mnitchev) [inline pkg/registrar/transitgateway.go:783]:
```suggestion
			{Name: aws.String("tag:" + subnetTGWAttachementsLabel), Values: []string{"true"}},
```

File: pkg/registrar/transitgateway.go [hunk 2]
```diff
@@ -32,6 +32,9 @@ const (
 	// when it is referenced. We're setting the max here to 45 for now so we stay below the
 	// default "Routes per route table" quota of 50.
 	PREFIX_LIST_MAX_ENTRIES = 45
+
+	subnetTGWAttachementslabel = "subnet.giantswarm.io/tgw-attachments"
```
Comment (mnitchev) [inline pkg/registrar/transitgateway.go:36]:
```suggestion
	subnetTGWAttachementsLabel = "subnet.giantswarm.io/tgw-attachments"
```

File: pkg/registrar/transitgateway.go [hunk 3]
```diff
@@ -771,6 +775,36 @@ func (r *TransitGateway) removeFromPrefixList(ctx context.Context, awsCluster *c
...
+			{Name: aws.String("tag:" + subnetTGWAttachementslabel), Values: []string{"true"}},
+			{Name: aws.String("tag:" + subnetRoleLabel), Values: []string{"private"}},
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if output == nil || len(output.Subnets) == 0 {
+		for _, s := range getPrivateSubnetsByAZ(awsCluster.Spec.NetworkSpec.Subnets) {
```
Comment (mnitchev) [inline pkg/registrar/transitgateway.go:792]: This function is used only here and we never really use the map for anything. I'd just make it return the desired slice of subnet IDs.

**Review by mnitchev — APPROVED** [review]
Not blocking. Just a few nitpicks

File: pkg/registrar/transitgateway.go [hunk 4]
```diff
@@ -771,6 +775,36 @@ func (r *TransitGateway) removeFromPrefixList(ctx context.Context, awsCluster *c
...
+			{Name: aws.String("tag:" + capa.NameKubernetesAWSCloudProviderPrefix + awsCluster.Name), Values: []string{"owned", "shared"}},
+			{Name: aws.String("tag:" + subnetTGWAttachementslabel), Values: []string{"true"}},
+			{Name: aws.String("tag:" + subnetRoleLabel), Values: []string{"private"}},
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if output == nil || len(output.Subnets) == 0 {
```
Comment (fiunchinho) [inline pkg/registrar/transitgateway.go:791]: can `output` ever be `nil` if `err != nil`?

Comment (bdehri) [reply pkg/registrar/transitgateway.go:791]: yes, if there are no matching subnets, describe returns empty output without any errors.

File: pkg/registrar/transitgateway.go [hunk 5]
```diff
@@ -771,6 +775,36 @@ func (r *TransitGateway) removeFromPrefixList(ctx context.Context, awsCluster *c
 	return nil
 }
 
+func (r *TransitGateway) getTGWGAttachmentSubnetsOrDefault(ctx context.Context, awsCluster *capa.AWSCluster) ([]string, error) {
```
Comment (fiunchinho) [inline pkg/registrar/transitgateway.go:778]: nitpick: I miss a func comment explaining a bit what this function does, what's the default and so on.

File: pkg/aws/ec2Client.go [hunk 6]
```diff
@@ -164,3 +164,11 @@ func (e *EC2Client) GetManagedPrefixListEntries(ctx context.Context, params *ec2
 	}
 	return client.GetManagedPrefixListEntries(ctx, params, optFns...)
 }
+
+func (e *EC2Client) DescribeSubnets(ctx context.Context, params *ec2.DescribeSubnetsInput, optFns ...func(*ec2.Options)) (*ec2.DescribeSubnetsOutput, error) {
```
Comment (fiunchinho) [inline pkg/aws/ec2Client.go:168]: @mnitchev are you changing this client in your refactoring? I don't see the point of having this wrapper around the sdk client at all.

Comment (mnitchev) [reply pkg/aws/ec2Client.go:168]: No I'm not. I'm creating a separate client for RAM

Comment (fiunchinho) [reply pkg/aws/ec2Client.go:168]: We can change it in a future PR, it's fine for this PR.

**Review by calvix — APPROVED** [review]

**Review by fiunchinho — APPROVED** [review]

**Pull request merged**